### PR TITLE
Added Async interface for async await development

### PR DESCRIPTION
**Pull request by amiryonatan**

### CODE REVIEW

File: StatsdClient/Statsd.cs [hunk 1]
```diff
@@ -1,199 +1,277 @@
...
-    }
+        /// <summary>
+        /// Log a timing / latency
+        /// </summary>
+        /// <param name="name">The metric name.</param>
+        /// <param name="milliseconds">The duration, in milliseconds, for this metric.</param>
+        public async Task LogTimingAsync(string name, long milliseconds)
+        {
+            await LogTimingAsync(name, (int)milliseconds);
+        }
```
Comment (niemyjski) [inline StatsdClient/Statsd.cs]: Why are you casting this to int?

File: StatsdClient/IOutputChannel.cs [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
     /// </summary>
-    void Send(string line);
-  }
-}
+    public interface IOutputChannel
+    {
+        /// <summary>
+        /// Sends a line of stats data to the server.
+        /// </summary>
+        void Send(string line);
```
Comment (niemyjski) [inline StatsdClient/IOutputChannel.cs]: I don't think this should be in the interface... Send(string) should just be an extension method.. It makes it easier to implementation and then sync method will never change and should call SendAsync().Wait()

File: StatsdClient/IStatsd.cs [hunk 3]
```diff
@@ -1,39 +1,104 @@
...
-  }
-}
+    public interface IStatsd
+    {
+        /// <summary>
+        /// Log a counter.
+        /// </summary>
+        /// <param name="name">The metric name.</param>
+        /// <param name="count">The counter value (defaults to 1).</param>
+        void LogCount(string name, int count = 1);
```
Comment (niemyjski) [inline StatsdClient/IStatsd.cs]: All sync methods should be an extension method.

File: StatsdClient/NullOutputChannel.cs [hunk 4]
```diff
@@ -1,15 +1,17 @@
...
 
 namespace StatsdClient
 {
-  internal sealed class NullOutputChannel : IOutputChannel
-  {
-    public void Send(string line)
+    internal sealed class NullOutputChannel : IOutputChannel
     {
-      // noop
+        public void Send(string line)
```
Comment (niemyjski) [inline StatsdClient/NullOutputChannel.cs]: All sync methods should be an extension methods.

File: StatsdClient/Statsd.cs [hunk 5]
```diff
@@ -1,199 +1,277 @@
...
-    public void LogTiming(string name, int milliseconds)
-    {
-      SendMetric(MetricType.TIMING, name, _prefix, milliseconds);
-    }
+        /// <summary>
+        /// Log a counter.
+        /// </summary>
+        /// <param name="name">The metric name.</param>
+        /// <param name="count">The counter value (defaults to 1).</param>
+        public void LogCount(string name, int count = 1)
```
Comment (niemyjski) [inline StatsdClient/Statsd.cs]: All sync methods should be an extension methods.

File: StatsdClient/TcpOutputChannel.cs [hunk 6]
```diff
@@ -1,85 +1,146 @@
...
-        if ( attemptsLeft > 0 )
+
+        public async Task SendAsync(string line)
         {
-          SendWithRetry( line, --attemptsLeft );
+            await SendWithRetryAsync(line, _reconnectEnabled ? _retryAttempts - 1 : 0);
         }
-        else
+
+        private void SendWithRetry(string line, int attemptsLeft)
```
Comment (niemyjski) [inline StatsdClient/TcpOutputChannel.cs]: There should only be an async version.

File: StatsdClient/TcpOutputChannel.cs [hunk 7]
```diff
@@ -1,85 +1,146 @@
...
-          _tcpClient.Connect(_host, _port);
-          _stream = _tcpClient.GetStream();
+            if (!_tcpClient.Connected)
+            {
+                using (await _asyncLock.LockAsync())
+                {
+                    if (!_tcpClient.Connected)
+                    {
+                        await _tcpClient.ConnectAsync(_host, _port);
+                        _stream = _tcpClient.GetStream();
```
Comment (niemyjski) [inline StatsdClient/TcpOutputChannel.cs]: Is there a GetStreamAsync?

File: StatsdClient/UdpOutputChannel.cs [hunk 8]
```diff
@@ -1,34 +1,46 @@
...
+            if (!IPAddress.TryParse(hostOrIPAddress, out ipAddress))
+            {
+                // Convert to ipv4 address
+                ipAddress = Dns.GetHostAddresses(hostOrIPAddress).First(p => p.AddressFamily == AddressFamily.InterNetwork);
+            }
+            _udpClient = new UdpClient();
+            _udpClient.Connect(ipAddress, port);
+        }
+
+        public void Send(string line)
```
Comment (niemyjski) [inline StatsdClient/UdpOutputChannel.cs]: Should only have async method.

**Comment by niemyjski** [comment]
Can you please review these changes. I think it would make it a lot cleaner and would be a great pr :)

**Comment by niemyjski** [comment]
@amiryonatan Can you please ensure you have the latest downstream so I can automatically merge the latest code in and if possible make the requested changes.

File: StatsdClient/Statsd.cs (hunk 1 above)
Comment (amiryonatan) [inline StatsdClient/Statsd.cs]: Because that's exactly how you did it in the sync implementation.

File: StatsdClient/TcpOutputChannel.cs (hunk 7 above)
Comment (amiryonatan) [inline StatsdClient/TcpOutputChannel.cs]: no

**Comment by amiryonatan** [comment]
Hey,

I pulled the latest version, and fixed all the remarks mentioned above.
I'm not sure why it says there are conflicts and I don't know how to see them.

Thanks

**Comment by niemyjski** [comment]
Thanks, when I have conflicts I use https://code.google.com/p/tortoisegit/ to resolve conflicts. It's a nice easy to use ui.

File: StatsdClient/AsyncLock.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace StatsdClient
+{
+    internal sealed class AsyncLock
```
Comment (niemyjski) [inline StatsdClient/AsyncLock.cs]: I almost wonder if we should be using the async lock from: https://github.com/StephenCleary/AsyncEx The only downside is now that you have to have a dependency on that.. As long as this lock works I think we should be fine. Just wish it was tested.

**Comment by niemyjski** [comment]
Your changes look good thanks! Is there any chance you can resolve the conflicts and then I'll merge this pull request.

**Pull request closed without merging**